### ISSUE TITLE
Streaming attachments event with `null` mail argument

### ISSUE THREAD

**Issue by stockholmux** [andris9/mailparser#112]
I'm streaming attachments and I've attached a function to the `attachment` event with mailparser version 0.4.8. Most of my emails are coming out with two arguments as expected(`attachment` and `mail`). However, I have a situation where the second argument (`mail`) is being returned as null. 

I've tracked it down to the `MailParser.prototype._processStateHeader` in [mailparser.js on line 334](https://github.com/andris9/mailparser/blob/master/lib/mailparser.js#L334). In the case where I'm getting `null` returned for the `mail` argument of that event, the `this._currentNode.parentNode` is `null` obviously. I inserted a `console.log` for the values of the `this._currentNode` and here is the result (sorry for the redactions, but it's a private project):

```
{ parentNode: null,
  headers:
   [ { key: 'return-path', value: '<[redacted]>' },
     { key: 'envelope-to', value: '[redacted]' },
     { key: 'delivery-date',
       value: 'Tue, 24 Feb 2015 01:55:02 -0600' },
     { key: 'received',
       value: 'from [redacted] ([[redacted]]:37418 helo=smtprelay.[redacted]) by [redacted] with esmtp (Exim 4.82) (envelope-from <[redacted]>) id 1YQAKb-0006Ib-Q5 for [redacted]; Tue, 24 Feb 2015 01:55:02 -0600' },
     { key: 'received',
       value: 'from [redacted] (unknown [[redacted]]) by[redacted] (Postfix) with ESMTP id 68BBA12BA13 for <[redacted]>; Tue, 24 Feb 2015 07:55:00 +0000 (UTC)' },
     { key: 'x-session-marker',
       value: '6B796C65406869676865726564636172656572732E6361' },
     { key: 'x-spam-summary',
       value: '2,0,0,,d41d8cd98f00b204,[redacted],:,RULES_HIT:355:379:541:871:988:989:1000:1185:1197:1260:1261:1311:1312:1313:1314:1345:1381:1514:1515:1516:1519:1522:1595:1596:1692:1730:1747:1777:1792:2392:2393:2559:2562:3138:3139:3140:3141:3142:3143:3554:3876:3877:4093:4512:6114:6351:6506:6642:6747:6748:7281:7652:9911:10004:10848:11604:11914:12517:12519:13895:21080:21088,0,RBL:none,CacheIP:none,Bayesian:0.5,0.5,0.5,Netcheck:none,DomainCache:0,MSF:not bulk,SPF:fn,MSBL:0,DNSBL:none,Custom_rules:0:0:0' },
     { key: 'x-he-tag', value: 'foot24_45902dbf51e1f' },
     { key: 'x-filterd-recvd-size', value: '1457370' },
     { key: 'received',
       value: 'from [[redacted]] (unknown [[redacted]]) (Authenticated sender: [redacted]) by omf06.[redacted] (Postfix) with ESMTPA for <[redacted]>; Tue, 24 Feb 2015 07:54:37 +0000 (UTC)' },
     { key: 'from', value: '[redacted] <[redacted]>' },
     { key: 'content-type',
       value: 'image/jpeg; x-unix-mode=0640; name="IMG_0055.JPG"' },
     { key: 'content-transfer-encoding', value: 'base64' },
     { key: 'subject', value: 'This is the subject' },
     { key: 'message-id',
       value: '<9CB9F7C3-F095-49B9-89EA-97DA66BC6026@[redacted]>' },
     { key: 'date', value: 'Tue, 24 Feb 2015 01:54:31 -0600' },
     { key: 'to', value: '[redacted]' },
     { key: 'mime-version',
       value: '1.0 (Mac OS X Mail 8.2 \\(2070.6\\))' },
     { key: 'content-disposition',
       value: 'inline; filename=IMG_0055.JPG' },
     { key: 'x-mailer', value: 'Apple Mail (2.2070.6)' } ],
  parsedHeaders:
   { 'return-path': '<[redacted]>',
     'envelope-to': '[redacted]',
     'delivery-date': 'Tue, 24 Feb 2015 01:55:02 -0600',
     received:
      [ 'from smtprelay0251.[redacted] ([216.40.44.251]:37418 helo=smtprelay.[redacted]) by [redacted] with esmtp (Exim 4.82) (envelope-from <[redacted]>) id 1YQAKb-0006Ib-Q5 for [redacted]; Tue, 24 Feb 2015 01:55:02 -0600',
        'from filter.[redacted] (unknown [216.40.38.60]) by smtprelay02.[redacted] (Postfix) with ESMTP id 68BBA12BA13 for <[redacted]>; Tue, 24 Feb 2015 07:55:00 +0000 (UTC)',
        'from [[redacted]] (unknown [[redacted]]) (Authenticated sender: [redacted]) by omf06.[redacted] (Postfix) with ESMTPA for <[redacted]>; Tue, 24 Feb 2015 07:54:37 +0000 (UTC)' ],
     'x-session-marker': '6B796C65406869676865726564636172656572732E6361',
     'x-spam-summary': '2,0,0,,d41d8cd98f00b204,[redacted],:,RULES_HIT:355:379:541:871:988:989:1000:1185:1197:1260:1261:1311:1312:1313:1314:1345:1381:1514:1515:1516:1519:1522:1595:1596:1692:1730:1747:1777:1792:2392:2393:2559:2562:3138:3139:3140:3141:3142:3143:3554:3876:3877:4093:4512:6114:6351:6506:6642:6747:6748:7281:7652:9911:10004:10848:11604:11914:12517:12519:13895:21080:21088,0,RBL:none,CacheIP:none,Bayesian:0.5,0.5,0.5,Netcheck:none,DomainCache:0,MSF:not bulk,SPF:fn,MSBL:0,DNSBL:none,Custom_rules:0:0:0',
     'x-he-tag': 'foot24_45902dbf51e1f',
     'x-filterd-recvd-size': '1457370',
     from: '[redacted] <[redacted]>',
     'content-type': 'image/jpeg; x-unix-mode=0640; name="IMG_0055.JPG"',
     'content-transfer-encoding': 'base64',
     subject: 'This is the subject',
     'message-id': '<9CB9F7C3-F095-49B9-89EA-97DA66BC6026@[redacted]>',
     date: 'Tue, 24 Feb 2015 01:54:31 -0600',
     to: '[redacted]',
     'mime-version': '1.0 (Mac OS X Mail 8.2 \\(2070.6\\))',
     'content-disposition': 'inline; filename=IMG_0055.JPG',
     'x-mailer': 'Apple Mail (2.2070.6)' },
  meta:
   { contentType: 'image/jpeg',
     fileName: 'IMG_0055.JPG',
     transferEncoding: 'base64',
     messageId: '9CB9F7C3-F095-49B9-89EA-97DA66BC6026@[redacted]',
     date: Tue Feb 24 2015 01:54:31 GMT-0600 (CST),
     contentDisposition: 'inline',
     generatedFileName: 'IMG_0055.JPG',
     contentId: 'ede8591aad4031c20d1dc58a5342a10b@mailparser',
     stream:
      { domain: null,
        _events: {},
        _maxListeners: 10,
        writable: true,
        checksum: [Object],
        length: 0,
        current: '' } },
  childNodes: [],
  priority: 'normal',
  from: [ { address: '[redacted]', name: '[redacted]' } ],
  subject: 'This is the subject',
  messageId: '9CB9F7C3-F095-49B9-89EA-97DA66BC6026@[redacted]',
  to: [ { address: '[redacted]', name: '' } ],
  useMIME: true,
  attachment: true,
  stream:
   { domain: null,
     _events: {},
     _maxListeners: 10,
     writable: true,
     checksum: { _binding: {}, _options: undefined },
     length: 0,
     current: '' } }
```
Which looks like the mail object itself. So, I think I've solved the problem by changing line 334 to:

```
this.emit("attachment", attachment, this._currentNode.parentNode || this._currentNode);
``` 

Am I missing something or is this a fix for the situation?

**Comment by andris9** [comment]
Yes, this was a but. It occurred when the attachment was the root node itself. Fixed in v0.4.9

**Comment by stockholmux** [comment]
Glad I could help.
